tsi1: test timeout (deadlock?) related to lifecycle.Resource.Close

Here was the first stack trace I had, from a full `go test ./...` on WIP based on master at 799edb87bcfdfd0db347ec7d58502b43c4e642eb.

```
SIGQUIT: quit
PC=0x7fff5dc44a1e m=0 sigcode=0

goroutine 0 [idle]:
runtime.pthread_cond_wait(0x1c28ae0, 0x1c28aa0, 0x7ffe00000000)
	/usr/local/Cellar/go/1.11.4/libexec/src/runtime/sys_darwin.go:302 +0x51
runtime.semasleep(0xffffffffffffffff, 0xc0)
	/usr/local/Cellar/go/1.11.4/libexec/src/runtime/os_darwin.go:63 +0x85
runtime.notesleep(0x1c288a0)
	/usr/local/Cellar/go/1.11.4/libexec/src/runtime/lock_sema.go:167 +0xe3
runtime.stopm()
	/usr/local/Cellar/go/1.11.4/libexec/src/runtime/proc.go:2016 +0xe3
runtime.findrunnable(0xc00002e500, 0x0)
	/usr/local/Cellar/go/1.11.4/libexec/src/runtime/proc.go:2487 +0x4dc
runtime.schedule()
	/usr/local/Cellar/go/1.11.4/libexec/src/runtime/proc.go:2613 +0x13a
runtime.park_m(0xc000001080)
	/usr/local/Cellar/go/1.11.4/libexec/src/runtime/proc.go:2676 +0xae
runtime.mcall(0x105b9db)
	/usr/local/Cellar/go/1.11.4/libexec/src/runtime/asm_amd64.s:299 +0x5b

goroutine 1 [chan receive, 9 minutes]:
testing.(*T).Run(0xc00013e600, 0x16f942f, 0x25, 0x170ab88, 0x107b701)
	/usr/local/Cellar/go/1.11.4/libexec/src/testing/testing.go:879 +0x383
testing.runTests.func1(0xc000190000)
	/usr/local/Cellar/go/1.11.4/libexec/src/testing/testing.go:1119 +0x78
testing.tRunner(0xc000190000, 0xc0000c7e08)
	/usr/local/Cellar/go/1.11.4/libexec/src/testing/testing.go:827 +0xbf
testing.runTests(0xc00014cca0, 0x1c217e0, 0x2a, 0x2a, 0x100d7df)
	/usr/local/Cellar/go/1.11.4/libexec/src/testing/testing.go:1117 +0x2aa
testing.(*M).Run(0xc00018a000, 0x0)
	/usr/local/Cellar/go/1.11.4/libexec/src/testing/testing.go:1034 +0x165
main.main()
	_testmain.go:154 +0x13d

goroutine 19 [syscall, 9 minutes]:
os/signal.signal_recv(0x0)
	/usr/local/Cellar/go/1.11.4/libexec/src/runtime/sigqueue.go:139 +0x9f
os/signal.loop()
	/usr/local/Cellar/go/1.11.4/libexec/src/os/signal/signal_unix.go:23 +0x22
created by os/signal.init.0
	/usr/local/Cellar/go/1.11.4/libexec/src/os/signal/signal_unix.go:29 +0x41

goroutine 38 [chan receive, 9 minutes]:
testing.(*T).Parallel(0xc000190400)
	/usr/local/Cellar/go/1.11.4/libexec/src/testing/testing.go:732 +0x1fa
github.com/influxdata/influxdb/tsdb/tsi1.TestTagValueSeriesIDCache_ConcurrentGetPutDelete(0xc000190400)
	/Users/mr/go/src/github.com/influxdata/influxdb/tsdb/tsi1/cache_test.go:166 +0x50
testing.tRunner(0xc000190400, 0x170aa28)
	/usr/local/Cellar/go/1.11.4/libexec/src/testing/testing.go:827 +0xbf
created by testing.(*T).Run
	/usr/local/Cellar/go/1.11.4/libexec/src/testing/testing.go:878 +0x35c

goroutine 422 [semacquire, 9 minutes]:
sync.runtime_SemacquireMutex(0xc00013e520, 0x3f6100)
	/usr/local/Cellar/go/1.11.4/libexec/src/runtime/sema.go:71 +0x3d
sync.(*RWMutex).Lock(0xc00013e518)
	/usr/local/Cellar/go/1.11.4/libexec/src/sync/rwmutex.go:98 +0x74
github.com/influxdata/influxdb/pkg/lifecycle.(*Resource).Close(0xc00013e518)
	/Users/mr/go/src/github.com/influxdata/influxdb/pkg/lifecycle/resource.go:28 +0x2d
github.com/influxdata/influxdb/tsdb/tsi1.(*IndexFile).Close(0xc00013e500, 0x0, 0x0)
	/Users/mr/go/src/github.com/influxdata/influxdb/tsdb/tsi1/index_file.go:147 +0x35
github.com/influxdata/influxdb/tsdb/tsi1.(*Partition).close(0xc0005e0240, 0x170bac0, 0xc0005e0240)
	/Users/mr/go/src/github.com/influxdata/influxdb/tsdb/tsi1/partition.go:412 +0xaa
github.com/influxdata/influxdb/tsdb/tsi1.(*Partition).Close(0xc0005e0240, 0x0, 0x0)
	/Users/mr/go/src/github.com/influxdata/influxdb/tsdb/tsi1/partition.go:395 +0xb7
github.com/influxdata/influxdb/tsdb/tsi1.(*Index).Close(0xc0004e5110, 0x0, 0x0)
	/Users/mr/go/src/github.com/influxdata/influxdb/tsdb/tsi1/index.go:351 +0xba
github.com/influxdata/influxdb/tsdb/tsi1_test.(*Index).Reopen(0xc00000c260, 0xc00c5cb0e0, 0x16add20)
	/Users/mr/go/src/github.com/influxdata/influxdb/tsdb/tsi1/index_test.go:457 +0x32
github.com/influxdata/influxdb/tsdb/tsi1_test.TestIndex_MeasurementCardinalityStats.func4(0xc00013e200)
	/Users/mr/go/src/github.com/influxdata/influxdb/tsdb/tsi1/index_test.go:403 +0x50f
testing.tRunner(0xc00013e200, 0x170ab80)
	/usr/local/Cellar/go/1.11.4/libexec/src/testing/testing.go:827 +0xbf
created by testing.(*T).Run
	/usr/local/Cellar/go/1.11.4/libexec/src/testing/testing.go:878 +0x35c

goroutine 411 [chan receive, 9 minutes]:
testing.(*T).Run(0xc00013e200, 0x16e916f, 0x5, 0x170ab80, 0x52ba5f01)
	/usr/local/Cellar/go/1.11.4/libexec/src/testing/testing.go:879 +0x383
github.com/influxdata/influxdb/tsdb/tsi1_test.TestIndex_MeasurementCardinalityStats(0xc00013e600)
	/Users/mr/go/src/github.com/influxdata/influxdb/tsdb/tsi1/index_test.go:381 +0xdd
testing.tRunner(0xc00013e600, 0x170ab88)
	/usr/local/Cellar/go/1.11.4/libexec/src/testing/testing.go:827 +0xbf
created by testing.(*T).Run
	/usr/local/Cellar/go/1.11.4/libexec/src/testing/testing.go:878 +0x35c

goroutine 9000 [semacquire, 9 minutes]:
sync.runtime_SemacquireMutex(0xc0005e0244, 0xc005213e00)
	/usr/local/Cellar/go/1.11.4/libexec/src/runtime/sema.go:71 +0x3d
sync.(*Mutex).Lock(0xc0005e0240)
	/usr/local/Cellar/go/1.11.4/libexec/src/sync/mutex.go:134 +0xff
sync.(*RWMutex).Lock(0xc0005e0240)
	/usr/local/Cellar/go/1.11.4/libexec/src/sync/rwmutex.go:93 +0x2d
github.com/influxdata/influxdb/tsdb/tsi1.(*Partition).Compact(0xc0005e0240)
	/Users/mr/go/src/github.com/influxdata/influxdb/tsdb/tsi1/partition.go:934 +0x2d
github.com/influxdata/influxdb/tsdb/tsi1.(*Partition).compact.func1(0xc000254b10, 0x2, 0x2, 0xc0005e0240, 0xc000254b00, 0x2, 0x2, 0xc00054c2a0, 0x2)
	/Users/mr/go/src/github.com/influxdata/influxdb/tsdb/tsi1/partition.go:1030 +0x12a
created by github.com/influxdata/influxdb/tsdb/tsi1.(*Partition).compact
	/Users/mr/go/src/github.com/influxdata/influxdb/tsdb/tsi1/partition.go:1016 +0x6d8

rax    0x104
rbx    0x34300
rcx    0x7ffeefbff348
rdx    0x34300
rdi    0x1c28ae0
rsi    0x3430100034400
rbp    0x7ffeefbff3e0
rsp    0x7ffeefbff348
r8     0x0
r9     0x60
r10    0x0
r11    0x202
r12    0x1c28ae0
r13    0x7ffeefbff368
r14    0x1
r15    0x7fff96117380
rip    0x7fff5dc44a1e
rflags 0x203
cs     0x7
fs     0x0
gs     0x0
*** Test killed with quit: ran too long (10m0s).
FAIL	github.com/influxdata/influxdb/tsdb/tsi1	600.118s
```

Then I ran the tsi1 suite successfully, saw it took about 23s on my machine, so set it off to run in a loop until I reproed the deadlock again. That time, from a clean commit of master at 799edb87bcfdfd0db347ec7d58502b43c4e642eb, I ran again and the first ran failed with this stack trace:

```
$ go test -c ./tsdb/tsi1/ && while ./tsi1.test -test.count=1 -test.timeout=45s; do date; done
panic: test timed out after 45s

goroutine 205 [running]:
testing.(*M).startAlarm.func1()
	/usr/local/Cellar/go/1.11.4/libexec/src/testing/testing.go:1296 +0xfd
created by time.goFunc
	/usr/local/Cellar/go/1.11.4/libexec/src/time/sleep.go:172 +0x44

goroutine 1 [chan receive]:
testing.(*T).Run(0xc000195800, 0x16f4274, 0x1d, 0x170ab20, 0x107b701)
	/usr/local/Cellar/go/1.11.4/libexec/src/testing/testing.go:879 +0x383
testing.runTests.func1(0xc000194000)
	/usr/local/Cellar/go/1.11.4/libexec/src/testing/testing.go:1119 +0x78
testing.tRunner(0xc000194000, 0xc0000bbe08)
	/usr/local/Cellar/go/1.11.4/libexec/src/testing/testing.go:827 +0xbf
testing.runTests(0xc0001561a0, 0x1c217e0, 0x2a, 0x2a, 0x100d7df)
	/usr/local/Cellar/go/1.11.4/libexec/src/testing/testing.go:1117 +0x2aa
testing.(*M).Run(0xc000186000, 0x0)
	/usr/local/Cellar/go/1.11.4/libexec/src/testing/testing.go:1034 +0x165
main.main()
	_testmain.go:154 +0x13d

goroutine 20 [syscall]:
os/signal.signal_recv(0x0)
	/usr/local/Cellar/go/1.11.4/libexec/src/runtime/sigqueue.go:139 +0x9f
os/signal.loop()
	/usr/local/Cellar/go/1.11.4/libexec/src/os/signal/signal_unix.go:23 +0x22
created by os/signal.init.0
	/usr/local/Cellar/go/1.11.4/libexec/src/os/signal/signal_unix.go:29 +0x41

goroutine 38 [chan receive]:
testing.(*T).Parallel(0xc000194400)
	/usr/local/Cellar/go/1.11.4/libexec/src/testing/testing.go:732 +0x1fa
github.com/influxdata/influxdb/tsdb/tsi1.TestTagValueSeriesIDCache_ConcurrentGetPutDelete(0xc000194400)
	/Users/mr/go/src/github.com/influxdata/influxdb/tsdb/tsi1/cache_test.go:166 +0x50
testing.tRunner(0xc000194400, 0x170aa28)
	/usr/local/Cellar/go/1.11.4/libexec/src/testing/testing.go:827 +0xbf
created by testing.(*T).Run
	/usr/local/Cellar/go/1.11.4/libexec/src/testing/testing.go:878 +0x35c

goroutine 204 [semacquire]:
sync.runtime_SemacquireMutex(0xc000300fd0, 0xc000000000)
	/usr/local/Cellar/go/1.11.4/libexec/src/runtime/sema.go:71 +0x3d
sync.(*RWMutex).Lock(0xc000300fc8)
	/usr/local/Cellar/go/1.11.4/libexec/src/sync/rwmutex.go:98 +0x74
github.com/influxdata/influxdb/pkg/lifecycle.(*Resource).Close(0xc000300fc8)
	/Users/mr/go/src/github.com/influxdata/influxdb/pkg/lifecycle/resource.go:28 +0x2d
github.com/influxdata/influxdb/tsdb.(*SeriesFile).Close(0xc000300fc0, 0x0, 0x0)
	/Users/mr/go/src/github.com/influxdata/influxdb/tsdb/series_file.go:162 +0x75
runtime.Goexit()
	/usr/local/Cellar/go/1.11.4/libexec/src/runtime/panic.go:397 +0xef
testing.(*common).FailNow(0xc000195800)
	/usr/local/Cellar/go/1.11.4/libexec/src/testing/testing.go:590 +0x39
testing.(*common).Fatal(0xc000195800, 0xc00028df78, 0x1, 0x1)
	/usr/local/Cellar/go/1.11.4/libexec/src/testing/testing.go:628 +0x6f
github.com/influxdata/influxdb/tsdb/tsi1_test.TestGenerateIndexFile_Uvarint(0xc000195800)
	/Users/mr/go/src/github.com/influxdata/influxdb/tsdb/tsi1/index_file_test.go:68 +0x1cf
testing.tRunner(0xc000195800, 0x170ab20)
	/usr/local/Cellar/go/1.11.4/libexec/src/testing/testing.go:827 +0xbf
created by testing.(*T).Run
	/usr/local/Cellar/go/1.11.4/libexec/src/testing/testing.go:878 +0x35c
```

... which was a user error on my part, because that line in index_file_test.go is trying to open testdata which isn't a valid path from the root of influxdb. But let's hold on to that stack trace anyway, because that lock attempt during close seems to be related. https://github.com/influxdata/influxdb/blob/e86a7d0a79f97ade30e74faeadd82f2b515500ca/tsdb/tsi1/index_file_test.go#L64-L70

After cd-ing to to tsdb/tsi1 so the testdata directory lookup would succeed, I ran:

```
$ go test -c . && while ./tsi1.test -test.count=1 -test.timeout=45s; do date; done
8<------- passed 31 times and then the 32nd run failed --------->8
panic: test timed out after 45s

goroutine 9456 [running]:
testing.(*M).startAlarm.func1()
	/usr/local/Cellar/go/1.11.4/libexec/src/testing/testing.go:1296 +0xfd
created by time.goFunc
	/usr/local/Cellar/go/1.11.4/libexec/src/time/sleep.go:172 +0x44

goroutine 1 [chan receive]:
testing.(*T).Run(0xc000540100, 0x16f942f, 0x25, 0x170ab88, 0x107b701)
	/usr/local/Cellar/go/1.11.4/libexec/src/testing/testing.go:879 +0x383
testing.runTests.func1(0xc000196000)
	/usr/local/Cellar/go/1.11.4/libexec/src/testing/testing.go:1119 +0x78
testing.tRunner(0xc000196000, 0xc0000c1e08)
	/usr/local/Cellar/go/1.11.4/libexec/src/testing/testing.go:827 +0xbf
testing.runTests(0xc00014ec60, 0x1c217e0, 0x2a, 0x2a, 0x100d7df)
	/usr/local/Cellar/go/1.11.4/libexec/src/testing/testing.go:1117 +0x2aa
testing.(*M).Run(0xc000192000, 0x0)
	/usr/local/Cellar/go/1.11.4/libexec/src/testing/testing.go:1034 +0x165
main.main()
	_testmain.go:154 +0x13d

goroutine 19 [syscall]:
os/signal.signal_recv(0x0)
	/usr/local/Cellar/go/1.11.4/libexec/src/runtime/sigqueue.go:139 +0x9f
os/signal.loop()
	/usr/local/Cellar/go/1.11.4/libexec/src/os/signal/signal_unix.go:23 +0x22
created by os/signal.init.0
	/usr/local/Cellar/go/1.11.4/libexec/src/os/signal/signal_unix.go:29 +0x41

goroutine 39 [chan receive]:
testing.(*T).Parallel(0xc000196400)
	/usr/local/Cellar/go/1.11.4/libexec/src/testing/testing.go:732 +0x1fa
github.com/influxdata/influxdb/tsdb/tsi1.TestTagValueSeriesIDCache_ConcurrentGetPutDelete(0xc000196400)
	/Users/mr/go/src/github.com/influxdata/influxdb/tsdb/tsi1/cache_test.go:166 +0x50
testing.tRunner(0xc000196400, 0x170aa28)
	/usr/local/Cellar/go/1.11.4/libexec/src/testing/testing.go:827 +0xbf
created by testing.(*T).Run
	/usr/local/Cellar/go/1.11.4/libexec/src/testing/testing.go:878 +0x35c

goroutine 389 [chan receive]:
testing.(*T).Run(0xc000138400, 0x16e916f, 0x5, 0x170ab80, 0x3d19cc01)
	/usr/local/Cellar/go/1.11.4/libexec/src/testing/testing.go:879 +0x383
github.com/influxdata/influxdb/tsdb/tsi1_test.TestIndex_MeasurementCardinalityStats(0xc000540100)
	/Users/mr/go/src/github.com/influxdata/influxdb/tsdb/tsi1/index_test.go:381 +0xdd
testing.tRunner(0xc000540100, 0x170ab88)
	/usr/local/Cellar/go/1.11.4/libexec/src/testing/testing.go:827 +0xbf
created by testing.(*T).Run
	/usr/local/Cellar/go/1.11.4/libexec/src/testing/testing.go:878 +0x35c

goroutine 8508 [semacquire]:
sync.runtime_SemacquireMutex(0xc0002a86c4, 0xc010b8be00)
	/usr/local/Cellar/go/1.11.4/libexec/src/runtime/sema.go:71 +0x3d
sync.(*Mutex).Lock(0xc0002a86c0)
	/usr/local/Cellar/go/1.11.4/libexec/src/sync/mutex.go:134 +0xff
sync.(*RWMutex).Lock(0xc0002a86c0)
	/usr/local/Cellar/go/1.11.4/libexec/src/sync/rwmutex.go:93 +0x2d
github.com/influxdata/influxdb/tsdb/tsi1.(*Partition).Compact(0xc0002a86c0)
	/Users/mr/go/src/github.com/influxdata/influxdb/tsdb/tsi1/partition.go:934 +0x2d
github.com/influxdata/influxdb/tsdb/tsi1.(*Partition).compact.func1(0xc009d91ad0, 0x2, 0x2, 0xc0002a86c0, 0xc009d91ac0, 0x2, 0x2, 0xc000084360, 0x2)
	/Users/mr/go/src/github.com/influxdata/influxdb/tsdb/tsi1/partition.go:1030 +0x12a
created by github.com/influxdata/influxdb/tsdb/tsi1.(*Partition).compact
	/Users/mr/go/src/github.com/influxdata/influxdb/tsdb/tsi1/partition.go:1016 +0x6d8

goroutine 459 [semacquire]:
sync.runtime_SemacquireMutex(0xc000540220, 0x3f6100)
	/usr/local/Cellar/go/1.11.4/libexec/src/runtime/sema.go:71 +0x3d
sync.(*RWMutex).Lock(0xc000540218)
	/usr/local/Cellar/go/1.11.4/libexec/src/sync/rwmutex.go:98 +0x74
github.com/influxdata/influxdb/pkg/lifecycle.(*Resource).Close(0xc000540218)
	/Users/mr/go/src/github.com/influxdata/influxdb/pkg/lifecycle/resource.go:28 +0x2d
github.com/influxdata/influxdb/tsdb/tsi1.(*IndexFile).Close(0xc000540200, 0x0, 0x0)
	/Users/mr/go/src/github.com/influxdata/influxdb/tsdb/tsi1/index_file.go:147 +0x35
github.com/influxdata/influxdb/tsdb/tsi1.(*Partition).close(0xc0002a86c0, 0x170bac0, 0xc0002a86c0)
	/Users/mr/go/src/github.com/influxdata/influxdb/tsdb/tsi1/partition.go:412 +0xaa
github.com/influxdata/influxdb/tsdb/tsi1.(*Partition).Close(0xc0002a86c0, 0x0, 0x0)
	/Users/mr/go/src/github.com/influxdata/influxdb/tsdb/tsi1/partition.go:395 +0xb7
github.com/influxdata/influxdb/tsdb/tsi1.(*Index).Close(0xc0004c6b60, 0x0, 0x0)
	/Users/mr/go/src/github.com/influxdata/influxdb/tsdb/tsi1/index.go:351 +0xba
github.com/influxdata/influxdb/tsdb/tsi1_test.(*Index).Reopen(0xc0000b0e00, 0xc002f3d4d0, 0x16add20)
	/Users/mr/go/src/github.com/influxdata/influxdb/tsdb/tsi1/index_test.go:457 +0x32
github.com/influxdata/influxdb/tsdb/tsi1_test.TestIndex_MeasurementCardinalityStats.func4(0xc000138400)
	/Users/mr/go/src/github.com/influxdata/influxdb/tsdb/tsi1/index_test.go:403 +0x50f
testing.tRunner(0xc000138400, 0x170ab80)
	/usr/local/Cellar/go/1.11.4/libexec/src/testing/testing.go:827 +0xbf
created by testing.(*T).Run
	/usr/local/Cellar/go/1.11.4/libexec/src/testing/testing.go:878 +0x35c
```

/cc @zeebo

--- a/pkg/lifecycle/resource.go
+++ b/pkg/lifecycle/resource.go
@@ -1,79 +1,89 @@
 package lifecycle
 
-import (
-	"sync"
-	"sync/atomic"
-	"unsafe"
-)
+import "sync"
 
 // Resource keeps track of references and has some compile time debug hooks
 // to help diagnose leaks. It keeps track of if it is open or not and allows
 // blocking until all references are released.
 type Resource struct {
-	sem  sync.RWMutex
-	open bool
+	stmu sync.Mutex     // protects state transitions
+	chmu sync.RWMutex   // protects channel mutations
+	ch   chan struct{}  // signals references to close
+	wg   sync.WaitGroup // counts outstanding references
 }
 
-// Open waits for any outstanding references, of which there should be none
-// and marks the reference counter as open.
+// Open marks the resource as open.
 func (res *Resource) Open() {
-	res.sem.Lock()
-	res.open = true
-	res.sem.Unlock()
+	res.stmu.Lock()
+	defer res.stmu.Unlock()
+
+	res.chmu.Lock()
+	res.ch = make(chan struct{})
+	res.chmu.Unlock()
 }
 
-// Close waits for any outstanding references and marks the reference counter
-// as closed, so that Acquire returns an error.
+// Close waits for any outstanding references and marks the resource as closed
+// so that Acquire returns an error.
 func (res *Resource) Close() {
-	res.sem.Lock()
-	res.open = false
-	res.sem.Unlock()
+	res.stmu.Lock()
+	defer res.stmu.Unlock()
+
+	res.chmu.Lock()
+	if res.ch != nil {
+		close(res.ch) // signal any references.
+		res.ch = nil  // stop future Acquires
+	}
+	res.chmu.Unlock()
+
+	res.wg.Wait() // wait for any acquired references
 }
 
-// Opened returns true if the resource is currently open.
+// Opened returns true if the resource is currently open. It may be immediately
+// false in the presence of concurrent Open and Close calls.
 func (res *Resource) Opened() bool {
-	res.sem.RLock()
-	open := res.open
-	res.sem.RUnlock()
-	return open
+	res.chmu.RLock()
+	opened := res.ch != nil
+	res.chmu.RUnlock()
+
+	return opened
 }
 
 // Acquire returns a Reference used to keep alive some resource.
 func (res *Resource) Acquire() (*Reference, error) {
-	res.sem.RLock()
-	if !res.open {
-		res.sem.RUnlock()
+	res.chmu.RLock()
+	defer res.chmu.RUnlock()
+
+	ch := res.ch
+	if ch == nil {
 		return nil, resourceClosed()
 	}
-	// RLock intentionally left open.
 
-	return live.track(&Reference{
-		res: res,
-		id:  0, // required because staticcheck
-	}), nil
+	res.wg.Add(1)
+	return live.track(&Reference{wg: &res.wg, ch: ch}), nil
 }
 
 // Reference is an open reference for some resource.
 type Reference struct {
-	res *Resource
-	id  uint64
+	once sync.Once
+	wg   *sync.WaitGroup
+	ch   <-chan struct{}
+	id   uint64
 }
 
+// Closing returns a channel that will be closed when the associated resource begins closing.
+func (ref *Reference) Closing() <-chan struct{} { return ref.ch }
+
 // Release causes the Reference to be freed. It is safe to call multiple times.
-func (r *Reference) Release() {
-	// Inline a sync.Once using the res pointer as the flag for if we have
-	// called unlock or not. This reduces the size of a ref.
-	addr := (*unsafe.Pointer)(unsafe.Pointer(&r.res))
-	old := atomic.LoadPointer(addr)
-	if old != nil && atomic.CompareAndSwapPointer(addr, old, nil) {
-		live.untrack(r)
-		(*Resource)(old).sem.RUnlock()
-	}
+func (ref *Reference) Release() {
+	ref.once.Do(func() {
+		live.untrack(ref)
+		ref.wg.Done()
+	})
 }
 
 // Close makes a Reference an io.Closer. It is safe to call multiple times.
-func (r *Reference) Close() error {
-	r.Release()
+func (ref *Reference) Close() error {
+	ref.Release()
 	return nil
 }
 
@@ -81,14 +91,14 @@ func (r *Reference) Close() error {
 type References []*Reference
 
 // Release releases all of the references. It is safe to call multiple times.
-func (rs References) Release() {
-	for _, r := range rs {
-		r.Release()
+func (refs References) Release() {
+	for _, ref := range refs {
+		ref.Release()
 	}
 }
 
-// Close makes references an io.Closer. It is safe to call multiple times.
-func (rs References) Close() error {
-	rs.Release()
+// Close makes References an io.Closer. It is safe to call multiple times.
+func (refs References) Close() error {
+	refs.Release()
 	return nil
 }

--- a/tsdb/tsi1/index.go
+++ b/tsdb/tsi1/index.go
@@ -318,12 +318,14 @@ func (i *Index) Compact() {
 	}
 }
 
+// EnableCompactions allows compactions to proceed again.
 func (i *Index) EnableCompactions() {
 	for _, p := range i.partitions {
 		p.EnableCompactions()
 	}
 }
 
+// DisableCompactions stops any ongoing compactions and waits for them to finish.
 func (i *Index) DisableCompactions() {
 	for _, p := range i.partitions {
 		p.DisableCompactions()

--- a/tsdb/tsi1/index_file.go
+++ b/tsdb/tsi1/index_file.go
@@ -127,10 +127,12 @@ func (f *IndexFile) Open() (err error) {
 
 	data, err := mmap.Map(f.Path(), 0)
 	if err != nil {
+		f.sfileref.Release()
 		return err
 	}
 
 	if err := f.UnmarshalBinary(data); err != nil {
+		f.sfileref.Release()
 		f.Close()
 		return err
 	}

--- a/tsdb/tsi1/index_test.go
+++ b/tsdb/tsi1/index_test.go
@@ -4,16 +4,21 @@ import (
 	"context"
 	"fmt"
 	"io/ioutil"
+	"math/rand"
 	"os"
 	"path/filepath"
 	"reflect"
 	"regexp"
+	"sync"
 	"testing"
+	"time"
 
 	"github.com/google/go-cmp/cmp"
 	"github.com/influxdata/influxdb/models"
 	"github.com/influxdata/influxdb/tsdb"
 	"github.com/influxdata/influxdb/tsdb/tsi1"
+	"go.uber.org/zap"
+	"go.uber.org/zap/zaptest"
 )
 
 // Bloom filter settings used in tests.
@@ -326,6 +331,8 @@ func TestIndex_DiskSizeBytes(t *testing.T) {
 
 // Ensure index can returns measurement cardinality stats.
 func TestIndex_MeasurementCardinalityStats(t *testing.T) {
+	t.Parallel()
+
 	t.Run("Empty", func(t *testing.T) {
 		idx := MustOpenIndex(1, tsi1.NewConfig())
 		defer idx.Close()
@@ -408,6 +415,99 @@ func TestIndex_MeasurementCardinalityStats(t *testing.T) {
 	})
 }
 
+// Ensure index keeps the correct set of series even with concurrent compactions.
+func TestIndex_CompactionConsistency(t *testing.T) {
+	t.Parallel()
+
+	idx := NewIndex(tsi1.DefaultPartitionN, tsi1.NewConfig())
+	idx.WithLogger(zaptest.NewLogger(t, zaptest.Level(zap.DebugLevel)))
+	if err := idx.Open(); err != nil {
+		t.Fatal(err)
+	}
+	defer idx.Close()
+
+	// Set up some framework to track launched goroutines.
+	wg, done := new(sync.WaitGroup), make(chan struct{})
+	spawn := func(fn func()) {
+		wg.Add(1)
+		go func() {
+			for {
+				select {
+				case <-done:
+					wg.Done()
+					return
+				default:
+					fn()
+				}
+			}
+		}()
+	}
+
+	// Spawn a goroutine to constantly ask the index to compact.
+	spawn(func() { idx.Compact() })
+
+	// Issue a number of writes and deletes for a while.
+	expected, operations := make(map[string]struct{}), []string(nil)
+	spawn(func() {
+		var err error
+		if len(expected) > 0 && rand.Intn(5) == 0 {
+			for m := range expected {
+				err = idx.DropMeasurement([]byte(m))
+				operations = append(operations, "delete: "+m)
+				delete(expected, m)
+				break
+			}
+		} else {
+			m := []byte(fmt.Sprintf("m%d", rand.Int()))
+			s := make([]Series, 100)
+			for i := range s {
+				s[i] = Series{Name: m, Tags: models.NewTags(map[string]string{fmt.Sprintf("t%d", i): "v"})}
+			}
+			err = idx.CreateSeriesSliceIfNotExists(s)
+			operations = append(operations, "add: "+string(m))
+			expected[string(m)] = struct{}{}
+		}
+		if err != nil {
+			t.Error(err)
+		}
+	})
+
+	// Let them run for a while and then wait.
+	time.Sleep(10 * time.Second)
+	close(done)
+	wg.Wait()
+
+	t.Log("expect", len(expected), "measurements after", len(operations), "operations")
+	for _, op := range operations {
+		t.Log(op)
+	}
+
+	for m := range expected {
+		if v, err := idx.MeasurementExists([]byte(m)); err != nil {
+			t.Fatal(err)
+		} else if !v {
+			t.Fatal("expected", m)
+		}
+	}
+
+	miter, err := idx.MeasurementIterator()
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer miter.Close()
+
+	for {
+		m, err := miter.Next()
+		if err != nil {
+			t.Fatal(err)
+		} else if m == nil {
+			break
+		} else if _, ok := expected[string(m)]; !ok {
+			t.Fatal("unexpected", string(m))
+		}
+	}
+}
+
 func BenchmarkIndex_ComputeMeasurementCardinalityStats(b *testing.B) {
 	idx := MustOpenIndex(1, tsi1.NewConfig())
 	defer idx.Close()
@@ -443,7 +543,7 @@ type Index struct {
 // NewIndex returns a new instance of Index at a temporary path.
 func NewIndex(partitionN uint64, c tsi1.Config) *Index {
 	idx := &Index{
-		Config:     tsi1.NewConfig(),
+		Config:     c,
 		SeriesFile: NewSeriesFile(),
 	}
 	idx.Index = tsi1.NewIndex(idx.SeriesFile.SeriesFile, idx.Config, tsi1.WithPath(MustTempDir()))

--- a/tsdb/tsi1/partition.go
+++ b/tsdb/tsi1/partition.go
@@ -41,8 +41,12 @@ const (
 
 // Partition represents a collection of layered index files and WAL.
 type Partition struct {
-	mu     sync.RWMutex
-	opened bool
+	// The rule to ensure no deadlocks, no resource leaks, and no use after close
+	// is that if the partition launches a goroutine, it must acquire a reference
+	// to itself first and releases it only after it has done all of its use of mu.
+	mu    sync.RWMutex
+	resmu sync.Mutex // protects res Open and Close
+	res   lifecycle.Resource
 
 	sfile    *tsdb.SeriesFile     // series lookup file
 	sfileref *lifecycle.Reference // reference to series lookup file
@@ -55,6 +59,7 @@ type Partition struct {
 	// NOTE: Does not include active log file stats.
 	stats MeasurementCardinalityStats
 
+	// Running statistics
 	tracker *partitionTracker
 
 	// Fast series lookup of series IDs in the series file that have been present
@@ -62,13 +67,10 @@ type Partition struct {
 	seriesIDSet *tsdb.SeriesIDSet
 
 	// Compaction management
-	levels          []CompactionLevel // compaction levels
-	levelCompacting []bool            // level compaction status
-
-	// Close management.
-	once    sync.Once
-	closing chan struct{} // closing is used to inform iterators the partition is closing.
-	wg      sync.WaitGroup
+	levels              []CompactionLevel // compaction levels
+	levelCompacting     []bool            // level compaction status
+	compactionsDisabled int               // counter of disables
+	compactionsWG       sync.WaitGroup
 
 	// Directory of the Partition's index files.
 	path string
@@ -78,10 +80,6 @@ type Partition struct {
 	MaxLogFileSize int64
 	nosync         bool // when true, flushing and syncing of LogFile will be disabled.
 	logbufferSize  int  // the LogFile's buffer is set to this value.
-
-	// Frequency of compaction checks.
-	compactionInterrupt chan struct{}
-	compactionsDisabled int
 
 	logger *zap.Logger
 
@@ -95,15 +93,11 @@ type Partition struct {
 // NewPartition returns a new instance of Partition.
 func NewPartition(sfile *tsdb.SeriesFile, path string) *Partition {
 	partition := &Partition{
-		closing:     make(chan struct{}),
 		path:        path,
 		sfile:       sfile,
 		seriesIDSet: tsdb.NewSeriesIDSet(),
 
 		MaxLogFileSize: DefaultMaxIndexLogFileSize,
-
-		// compactionEnabled: true,
-		compactionInterrupt: make(chan struct{}),
 
 		logger:  zap.NewNop(),
 		version: Version,
@@ -117,12 +111,17 @@ func NewPartition(sfile *tsdb.SeriesFile, path string) *Partition {
 // bytes estimates the memory footprint of this Partition, in bytes.
 func (p *Partition) bytes() int {
 	var b int
-	b += 24 // mu RWMutex is 24 bytes
-	b += int(unsafe.Sizeof(p.opened))
-	// Do not count SeriesFile because it belongs to the code that constructed this Partition.
+	b += int(unsafe.Sizeof(p.mu))
+	b += int(unsafe.Sizeof(p.resmu))
+	b += int(unsafe.Sizeof(p.res))
+	// Do not count SeriesFile contents because it belongs to the code that constructed this Partition.
+	b += int(unsafe.Sizeof(p.sfile))
+	b += int(unsafe.Sizeof(p.sfileref))
 	b += int(unsafe.Sizeof(p.activeLogFile)) + p.activeLogFile.bytes()
 	b += int(unsafe.Sizeof(p.fileSet)) + p.fileSet.bytes()
 	b += int(unsafe.Sizeof(p.seq))
+	b += int(unsafe.Sizeof(p.stats))
+	b += int(unsafe.Sizeof(p.tracker))
 	b += int(unsafe.Sizeof(p.seriesIDSet)) + p.seriesIDSet.Bytes()
 	b += int(unsafe.Sizeof(p.levels))
 	for _, level := range p.levels {
@@ -132,14 +131,12 @@ func (p *Partition) bytes() int {
 	for _, levelCompacting := range p.levelCompacting {
 		b += int(unsafe.Sizeof(levelCompacting))
 	}
-	b += 12 // once sync.Once is 12 bytes
-	b += int(unsafe.Sizeof(p.closing))
-	b += 16 // wg sync.WaitGroup is 16 bytes
+	b += int(unsafe.Sizeof(p.compactionsDisabled))
 	b += int(unsafe.Sizeof(p.path)) + len(p.path)
 	b += int(unsafe.Sizeof(p.id)) + len(p.id)
 	b += int(unsafe.Sizeof(p.MaxLogFileSize))
-	b += int(unsafe.Sizeof(p.compactionInterrupt))
-	b += int(unsafe.Sizeof(p.compactionsDisabled))
+	b += int(unsafe.Sizeof(p.nosync))
+	b += int(unsafe.Sizeof(p.logbufferSize))
 	b += int(unsafe.Sizeof(p.logger))
 	b += int(unsafe.Sizeof(p.manifestSize))
 	b += int(unsafe.Sizeof(p.version))
@@ -152,10 +149,10 @@ var ErrIncompatibleVersion = errors.New("incompatible tsi1 index MANIFEST")
 
 // Open opens the partition.
 func (p *Partition) Open() (err error) {
-	p.mu.Lock()
-	defer p.mu.Unlock()
+	p.resmu.Lock()
+	defer p.resmu.Unlock()
 
-	if p.opened {
+	if p.res.Opened() {
 		return errors.New("index partition already open")
 	}
 
@@ -165,7 +162,6 @@ func (p *Partition) Open() (err error) {
 		return err
 	}
 
-	p.closing = make(chan struct{})
 	defer func() {
 		if err != nil {
 			p.close()
@@ -250,6 +246,9 @@ func (p *Partition) Open() (err error) {
 	// Place the files in a file set.
 	p.fileSet, err = NewFileSet(p.sfile, files)
 	if err != nil {
+		for _, file := range files {
+			file.Close()
+		}
 		return err
 	}
 
@@ -281,7 +280,7 @@ func (p *Partition) Open() (err error) {
 	p.computeStats()
 
 	// Mark opened.
-	p.opened = true
+	p.res.Open()
 
 	// Send a compaction request on start up.
 	p.compact()
@@ -364,25 +363,17 @@ func (p *Partition) buildSeriesSet() error {
 	return nil
 }
 
-// Wait returns once outstanding compactions have finished.
-func (p *Partition) Wait() {
-	p.wg.Wait()
-}
-
-// Close closes the index.
+// Close closes the partition.
 func (p *Partition) Close() error {
-	// Wait for goroutines to finish outstanding compactions.
-	p.once.Do(func() {
-		if p.closing != nil {
-			close(p.closing)
-		}
-		if p.compactionInterrupt != nil {
-			close(p.compactionInterrupt)
-		}
-	})
-	p.wg.Wait()
+	p.resmu.Lock()
+	defer p.resmu.Unlock()
 
-	// Lock index.
+	// Close the resource and wait for any outstanding references.
+	p.res.Close()
+	p.compactionsWG.Wait()
+
+	// There are now no internal outstanding callers holding a reference
+	// so we can acquire this mutex to protect against external callers.
 	p.mu.Lock()
 	defer p.mu.Unlock()
 
@@ -411,17 +402,6 @@ func (p *Partition) close() error {
 	}
 
 	return err
-}
-
-// closing returns true if the partition is currently closing. It does not require
-// a lock so will always return to callers.
-func (p *Partition) isClosing() bool {
-	select {
-	case <-p.closing:
-		return true
-	default:
-		return false
-	}
 }
 
 // Path returns the path to the partition.
@@ -480,19 +460,10 @@ func (p *Partition) WithLogger(logger *zap.Logger) {
 // FileSet returns a copy of the current file set. You must call Release on it when
 // you are finished.
 func (p *Partition) FileSet() (*FileSet, error) {
-	// It's important to check p.closing with the read lock so that even if
-	// the information about if the partition being closed is immediately wrong
-	// after the select case decision, the rest of the close procedure cannot
-	// run until after we have constructed the file set.
 	p.mu.RLock()
-	defer p.mu.RUnlock()
-
-	select {
-	case <-p.closing:
-		return nil, errors.New("index is closing")
-	default:
-		return p.fileSet.Duplicate()
-	}
+	fs, err := p.fileSet.Duplicate()
+	p.mu.RUnlock()
+	return fs, err
 }
 
 // replaceFileSet is a helper to replace the file set of the partition. It releases
@@ -516,6 +487,7 @@ func (p *Partition) prependActiveLogFile() error {
 	// Prepend and generate new fileset.
 	fileSet, err := p.fileSet.PrependLogFile(f)
 	if err != nil {
+		f.Close()
 		return err
 	}
 
@@ -524,6 +496,7 @@ func (p *Partition) prependActiveLogFile() error {
 	if err != nil {
 		// TODO: Close index if write fails.
 		fileSet.Release()
+		f.Close()
 		return err
 	}
 
@@ -914,53 +887,41 @@ func (p *Partition) AssignShard(k string, shardID uint64)         {}
 func (p *Partition) Compact() {
 	p.mu.Lock()
 	defer p.mu.Unlock()
+
 	p.compact()
 }
 
+// DisableCompactions stops any compactions from starting until a call to EnableCompactions.
 func (p *Partition) DisableCompactions() {
 	p.mu.Lock()
 	defer p.mu.Unlock()
+
 	p.compactionsDisabled++
-
-	select {
-	case <-p.closing:
-		return
-	default:
-	}
-
-	if p.compactionsDisabled == 0 {
-		close(p.compactionInterrupt)
-		p.compactionInterrupt = make(chan struct{})
-	}
 }
 
+// EnableCompactions allows compactions to proceed again after a call to DisableCompactions.
 func (p *Partition) EnableCompactions() {
 	p.mu.Lock()
 	defer p.mu.Unlock()
 
-	// Already enabled?
-	if p.compactionsEnabled() {
-		return
-	}
 	p.compactionsDisabled--
 }
 
-func (p *Partition) compactionsEnabled() bool {
-	return p.compactionsDisabled == 0
-}
+// Wait will block until all compactions are finished. Must only be called while they
+// are disabled.
+func (p *Partition) Wait() { p.compactionsWG.Wait() }
 
 // compact compacts continguous groups of files that are not currently compacting.
 func (p *Partition) compact() {
-	if p.isClosing() {
-		return
-	} else if !p.compactionsEnabled() {
+	if p.compactionsDisabled > 0 {
+		p.logger.Error("Cannot start a compaction while disabled")
 		return
 	}
-	interrupt := p.compactionInterrupt
 
 	fs, err := p.fileSet.Duplicate()
 	if err != nil {
 		p.logger.Error("Attempt to compact while partition is closing", zap.Error(err))
+		return
 	}
 	defer fs.Release()
 
@@ -982,6 +943,13 @@ func (p *Partition) compact() {
 			files = files[len(files)-MaxIndexMergeCount:]
 		}
 
+		// We intend to do a compaction. Acquire a resource to do so.
+		ref, err := p.res.Acquire()
+		if err != nil {
+			p.logger.Error("Attempt to compact while partition is closing", zap.Error(err))
+			return
+		}
+
 		// Acquire references to the files to keep them alive through compaction.
 		frefs, err := IndexFiles(files).Acquire()
 		if err != nil {
@@ -993,19 +961,20 @@ func (p *Partition) compact() {
 		p.levelCompacting[level] = true
 
 		// Start compacting in a separate goroutine.
-		p.wg.Add(1)
+		p.compactionsWG.Add(1)
 		go func(level int) {
-			// Ensure file references are always released.
-			defer frefs.Release()
-
 			// Compact to a new level.
-			p.compactToLevel(files, frefs, level+1, interrupt)
+			p.compactToLevel(files, frefs, level+1, ref.Closing())
 
 			// Ensure compaction lock for the level is released.
 			p.mu.Lock()
 			p.levelCompacting[level] = false
 			p.mu.Unlock()
-			p.wg.Done()
+
+			// Ensure references are released.
+			frefs.Release()
+			ref.Release()
+			p.compactionsWG.Done()
 
 			// Check for new compactions
 			p.Compact()
@@ -1143,8 +1112,6 @@ func (p *Partition) compactToLevel(files []*IndexFile, frefs lifecycle.Reference
 	}
 }
 
-func (p *Partition) Rebuild() {}
-
 func (p *Partition) CheckLogFile() error {
 	// Check log file size under read lock.
 	p.mu.RLock()
@@ -1162,7 +1129,18 @@ func (p *Partition) CheckLogFile() error {
 }
 
 func (p *Partition) checkLogFile() error {
+	if p.compactionsDisabled > 0 {
+		return nil
+	}
+
+	// Acquire a reference to hold the partition open.
+	ref, err := p.res.Acquire()
+	if err != nil {
+		return err
+	}
+
 	if p.activeLogFile.Size() < p.MaxLogFileSize {
+		ref.Release()
 		return nil
 	}
 
@@ -1171,15 +1149,17 @@ func (p *Partition) checkLogFile() error {
 
 	// Open new log file and insert it into the first position.
 	if err := p.prependActiveLogFile(); err != nil {
+		ref.Release()
 		return err
 	}
 
 	// Begin compacting in a background goroutine.
-	p.wg.Add(1)
+	p.compactionsWG.Add(1)
 	go func() {
-		defer p.wg.Done()
-		p.compactLogFile(logFile)
-		p.Compact() // check for new compactions
+		p.compactLogFile(logFile, ref.Closing())
+		ref.Release()          // release our reference
+		p.compactionsWG.Done() // compaction is now complete
+		p.Compact()            // check for new compactions
 	}()
 
 	return nil
@@ -1188,11 +1168,7 @@ func (p *Partition) checkLogFile() error {
 // compactLogFile compacts f into a tsi file. The new file will share the
 // same identifier but will have a ".tsi" extension. Once the log file is
 // compacted then the manifest is updated and the log file is discarded.
-func (p *Partition) compactLogFile(logFile *LogFile) {
-	if p.isClosing() {
-		return
-	}
-
+func (p *Partition) compactLogFile(logFile *LogFile, interrupt <-chan struct{}) {
 	defer func() {
 		p.mu.RLock()
 		defer p.mu.RUnlock()
@@ -1200,10 +1176,6 @@ func (p *Partition) compactLogFile(logFile *LogFile) {
 		p.tracker.SetFiles(uint64(len(p.fileSet.LogFiles())), "log")
 		p.tracker.SetDiskSize(uint64(p.fileSet.Size()))
 	}()
-
-	p.mu.Lock()
-	interrupt := p.compactionInterrupt
-	p.mu.Unlock()
 
 	start := time.Now()
 

--- a/tsdb/tsi1/partition_test.go
+++ b/tsdb/tsi1/partition_test.go
@@ -12,8 +12,6 @@ import (
 )
 
 func TestPartition_Open(t *testing.T) {
-	t.Parallel() // There's a bit of IO in this test.
-
 	sfile := MustOpenSeriesFile()
 	defer sfile.Close()
 
@@ -26,15 +24,20 @@ func TestPartition_Open(t *testing.T) {
 
 		fs, err := p.FileSet()
 		if err != nil {
+			p.Close()
 			t.Fatal(err)
 		}
 		defer fs.Release()
 
 		// Check version set appropriately.
 		if got, exp := p.Manifest(fs).Version, 1; got != exp {
+			p.Close()
 			t.Fatalf("got index version %d, expected %d", got, exp)
 		}
 	})
+	if t.Failed() {
+		return
+	}
 
 	// Reopening an open index should return an error.
 	t.Run("reopen open index", func(t *testing.T) {
@@ -45,6 +48,9 @@ func TestPartition_Open(t *testing.T) {
 		}
 		p.Close()
 	})
+	if t.Failed() {
+		return
+	}
 
 	// Opening an incompatible index should return an error.
 	incompatibleVersions := []int{-1, 0, 2}
@@ -75,6 +81,9 @@ func TestPartition_Open(t *testing.T) {
 				t.Fatalf("got error %v, expected %v", err, tsi1.ErrIncompatibleVersion)
 			}
 		})
+		if t.Failed() {
+			return
+		}
 	}
 }
 

